src/Mobs/PathFinder.cpp

```diff
@@ -0,0 +1,237 @@
+
+
+#include "Globals.h"
+#include "PathFinder.h"
+#include "../Chunk.h"
+
+#define WAYPOINT_RADIUS 0.25
+cPathFinder::cPathFinder(double a_MobWidth, double a_MobHeight) :
+	m_Path(),
+	m_GiveUpCounter(0),
+	m_NotFoundCooldown(0)
+{
+	m_Width = a_MobWidth;
+	m_Height = a_MobHeight;
+}
+
+// TODO fix monster heads
+// TODO why does PATH_NOT_FOUND even occur?
+// TODO fix holding shift on an edge issue (ensureProperDestination?)
+
+
+ePathFinderStatus cPathFinder::GetNextWayPoint(cChunk &a_Chunk, Vector3d a_Source, Vector3d a_Destination, Vector3d &a_OutputWaypoint)
+{
+	m_FinalDestination = a_Destination;
+	m_Source = a_Source;
+
+
+	// If a PATH_NOT_FOUND was recently returned, we rest for a few ticks.
+	if (m_NotFoundCooldown > 0)
+	{
+		m_NotFoundCooldown -= 1;
+		return ePathFinderStatus::PATH_NOT_FOUND;
+	}
+
+	// Tweak the destination. If something is wrong with the destination or the chunk, rest for a while.
+	if (!EnsureProperDestination(a_Chunk))
+	{
+		m_NotFoundCooldown = 20;
+		return ePathFinderStatus::PATH_NOT_FOUND;
+	}
+
+	// Rest is over. Prepare m_Path by calling ResetPathFinding.
+	if (m_NotFoundCooldown == 0)
+	{
+		m_NotFoundCooldown = -1;
+		ResetPathFinding(a_Chunk);
+	}
+
+	// If the mob is trying to go to where it currently is, then it's pointless.
+	if ((m_Source - m_FinalDestination).SqrLength() < WAYPOINT_RADIUS)
+	{
+		a_OutputWaypoint = m_Source;
+		return ePathFinderStatus::PATH_FOUND;
+	}
+
+	// If m_Path has not been initialized yet, initialize it.
+	if (!m_Path.IsValid())
+	{
+		ResetPathFinding(a_Chunk);
+	}
+
+	switch (m_Path.Step(a_Chunk))
```

Comment:
To prevent blocking for hard path calculations, the calculation is split into ticks. Each tick does a portion of the calculation. However, the word `tick` can be confused with Minecraft ticks, so I called it `step`.

Each call calculates a small part of the path, until the calculation is done.